waku/waku_filter_v2/subscriptions.nim

```diff
@@ -6,48 +6,161 @@ else:
+    if data.criteriaCount == 0:
+      return subscribedContentTopics
 
-  # TODO: for large maps, this can be optimized using a reverse index
-  for (peerId, criteria) in subscriptions.pairs():
-    if filterCriterion in criteria:
-      subscribedPeers.add(peerId)
+    for filterCriterion, subscribedPeers in s.subscriptions.mpairs:
+      if peerId in subscribedPeers:
+        subscribedContentTopics.add(filterCriterion)
 
-  subscribedPeers
+  return subscribedContentTopics
 
-proc removePeer*(subscriptions: var FilterSubscriptions, peerId: PeerID) =
+proc findSubscribedPeers*(s: var FilterSubscriptions, pubsubTopic: PubsubTopic, contentTopic: ContentTopic): seq[PeerID] =
+  let filterCriterion : FilterCriterion = (pubsubTopic, contentTopic)
+
+  var foundPeers : seq[PeerID] = @[]
+  # only peers subscribed to criteria and with legit subscription is counted
+  s.subscriptions.withValue(filterCriterion, peers):
+    for peer in peers[]:
+      if s.isSubscribed(peer):
+        foundPeers.add(peer)
+
+  return foundPeers
+
+proc removePeer*(s: var FilterSubscriptions, peerId: PeerID) =
   ## Remove all subscriptions for a given peer
-  subscriptions.del(peerId)
+  s.peersSubscribed.del(peerId)
 
-proc removePeers*(subscriptions: var FilterSubscriptions, peerIds: seq[PeerID]) =
+proc removePeers*(s: var FilterSubscriptions, peerIds: seq[PeerID]) =
   ## Remove all subscriptions for a given list of peers
   for peerId in peerIds:
-    subscriptions.removePeer(peerId)
-
-proc containsAny*(criteria: FilterCriteria, otherCriteria: FilterCriteria): bool =
-  ## Check if a given pubsub topic is contained in a set of filter criteria
-  ## TODO: Better way to achieve this?
-  for criterion in otherCriteria:
-    if criterion in criteria:
-      return true
-  false
+    s.removePeer(peerId)
+
+proc cleanUp*(fs: var FilterSubscriptions) =
+  ## Remove all subscriptions for peers that have not been seen for a while
+  let now = Moment.now()
+  fs.peersSubscribed.keepItIf(now - val.lastSeen <= fs.subscriptionTimeout)
+
+  var filtersToRemove: seq[FilterCriterion] = @[]
+  for filterCriterion, subscribedPeers in fs.subscriptions.mpairs:
+    subscribedPeers.keepItIf(fs.isSubscribed(it)==true)
+
+  fs.subscriptions.keepItIf(val.len() > 0)
+
+proc refreshSubscription*(s: var FilterSubscriptions, peerId: PeerID) =
+  s.peersSubscribed.withValue(peerId, data):
+    data.lastSeen = Moment.now()
+
+proc addSubscription*(s: var FilterSubscriptions, peerId: PeerID, filterCriteria: FilterCriteria): Result[void, string] =
+  ## Add a subscription for a given peer
+  var peerData: ptr PeerData
+
+  s.peersSubscribed.withValue(peerId, data):
+    if data.criteriaCount + cast[uint](filterCriteria.len()) > s.maxCriteriaPerPeer:
+      return err("Max number of criteria for peer (" & $s.maxCriteriaPerPeer & ") reached")
+
+    data.lastSeen = Moment.now()
+    peerData = data
+
+  do:
+    ## not yet subscribed
+    if cast[uint](s.peersSubscribed.len()) >= s.maxPeers:
```

Comment:
There's a different usage in the equalities for both `max` values:
```nim
> s.maxCriteriaPerPeer  # line 121
>= s.maxPeers # line 129
```
Is this due to some particular reason? It feels like we should be using the same "method"